x/pki/handler_test.go

```diff
@@ -0,0 +1,1522 @@
+	require.Equal(t, testconstants.RootSubject, approvedCertificate.Subject)
+	require.Equal(t, testconstants.RootSubjectKeyID, approvedCertificate.SubjectKeyId)
+	require.Equal(t, testconstants.RootSerialNumber, approvedCertificate.SerialNumber)
+	require.True(t, approvedCertificate.IsRoot)
+	require.Empty(t, approvedCertificate.RootSubject)
+	require.Empty(t, approvedCertificate.RootSubjectKeyId)
+	require.Empty(t, approvedCertificate.Issuer)
+	require.Empty(t, approvedCertificate.AuthorityKeyId)
+
+	// check that unique certificate key is registered
+	require.True(t, setup.Keeper.IsUniqueCertificatePresent(
+		setup.Ctx, testconstants.RootIssuer, testconstants.RootSerialNumber))
+}
+
+func TestHandler_ApproveAddX509RootCert_ForUnknownProposedCertificate(t *testing.T) {
+	setup := Setup(t)
+
+	// approve
+	approveAddX509RootCert := types.NewMsgApproveAddX509RootCert(
+		setup.Trustee.String(), testconstants.RootSubject, testconstants.RootSubjectKeyID)
+	_, err := setup.Handler(setup.Ctx, approveAddX509RootCert)
+	require.Error(t, err)
+	require.True(t, types.ErrProposedCertificateDoesNotExist.Is(err))
+}
+
+func TestHandler_ApproveAddX509RootCert_ByNotTrustee(t *testing.T) {
+	setup := Setup(t)
+
+	// propose add x509 root certificate
+	proposeAddX509RootCert := types.NewMsgProposeAddX509RootCert(setup.Trustee.String(), testconstants.RootCertPem)
+	_, err := setup.Handler(setup.Ctx, proposeAddX509RootCert)
+	require.NoError(t, err)
+
+	for _, role := range []dclauthtypes.AccountRole{
+		dclauthtypes.Vendor,
+		dclauthtypes.TestHouse,
+		dclauthtypes.CertificationCenter,
+		dclauthtypes.NodeAdmin,
+	} {
+		accAddress := GenerateAccAddress()
+		setup.AddAccount(accAddress, []dclauthtypes.AccountRole{role})
+
+		// approve
+		approveAddX509RootCert := types.NewMsgApproveAddX509RootCert(
+			accAddress.String(), testconstants.RootSubject, testconstants.RootSubjectKeyID)
+		_, err = setup.Handler(setup.Ctx, approveAddX509RootCert)
+		require.Error(t, err)
+		require.True(t, sdkerrors.ErrUnauthorized.Is(err))
+	}
+}
+
+func TestHandler_ApproveAddX509RootCert_Twice(t *testing.T) {
+	setup := Setup(t)
+
+	// store account without Trustee role
+	accAddress := GenerateAccAddress()
+	setup.AddAccount(accAddress, []dclauthtypes.AccountRole{dclauthtypes.Vendor})
+
+	// propose add x509 root certificate
+	proposeAddX509RootCert := types.NewMsgProposeAddX509RootCert(accAddress.String(), testconstants.RootCertPem)
+	_, err := setup.Handler(setup.Ctx, proposeAddX509RootCert)
+	require.NoError(t, err)
+
+	// approve
+	approveAddX509RootCert := types.NewMsgApproveAddX509RootCert(
+		setup.Trustee.String(), testconstants.RootSubject, testconstants.RootSubjectKeyID)
+	_, err = setup.Handler(setup.Ctx, approveAddX509RootCert)
+	require.NoError(t, err)
+
+	// approve second time
+	_, err = setup.Handler(setup.Ctx, approveAddX509RootCert)
+	require.Error(t, err)
+	require.True(t, sdkerrors.ErrUnauthorized.Is(err))
```

Comment:
As I can see, it was implemented to raise ErrUnauthorized in legacy as well, but it looks weird. Probably better to create and raise a special error code in `pki/types/errors.go`, something like `ErrProposedCertificateAlreadyApproved`.